datafusion/src/datasource/object_store/local.rs

```diff
@@ -0,0 +1,173 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Object store that represents the Local File System.
+
+use std::fs::Metadata;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use futures::{stream, AsyncRead, StreamExt};
+
+use crate::datasource::object_store::{
+    FileMeta, FileMetaStream, ObjectReader, ObjectStore,
+};
+use crate::error::DataFusionError;
+use crate::error::Result;
+
+#[derive(Debug)]
+/// Local File System as Object Store.
+pub struct LocalFileSystem;
+
+#[async_trait]
+impl ObjectStore for LocalFileSystem {
+    async fn list(
+        &self,
+        prefix: &str,
+        _delimiter: Option<String>,
+    ) -> Result<FileMetaStream> {
+        list_all(prefix.to_owned()).await
```

Comment:
we can maybe mark this as `// TODO` as the delimiter is ignored 😉

Reply:
it seems to me that if you don't comply to the API, a TODO is appropriate even if nobody is using the feature yet 😃

Reply:
`PathMeta`?  `FilePathMeta`? 😂 I'm run out of naming

Reply:
naming is hard indeed :D 

how about `ListEntry` or `ListItem`? we can define it as an enum of `FileMeta` and prefix string to distinguish between regular objects and "folder" prefixes.

References:
S3 has prefixes returned in the `  <CommonPrefixes>` section: https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html
GCS has prefixes returned in response object's prefixes property: https://cloud.google.com/storage/docs/json_api/v1/objects/list#response
Azure blob is similar to S3 and has prefix strings returned in its `BlobPrefix` section: https://docs.microsoft.com/en-us/rest/api/storageservices/enumerating-blob-resources#DelimitedBlobList

Reply:
There might be value in having two list apis as well, one for listing objects without delimiters and returns stream of `FileMeta`, the other for listing "directories" with delimiters that could return either `FileMeta` or prefix strings.

I am thinking there are use-case where we only need to perform listing without delimiters. For example after we have partition "folders" fully discovered, we will just use list api call to get the final list of objects using partition prefixes. Returning a stream of `FileMeta` directly in this case could avoid enum matching overhead on the call site.

Reply:
FWIW in IOx we went back and forth on this as well. There we have one LIST API https://github.com/influxdata/influxdb_iox/blob/main/object_store/src/lib.rs#L95-L98 that takes an`ObjectStorePath` and the path itself can distinguish between "is a (logical)  directory" or "is a filename" : https://github.com/influxdata/influxdb_iox/blob/main/object_store/src/path.rs#L30-L42

Reply:
@alamb that's interesting, do you have any opinion on the design here?

Since list without delimiter always returns object keys, not logical directory keys, I feel like returning `FileMeta` in list call provides a better more statically typed interface for callers.

In IOx, do you remember what was the reason for going back to `ObjectStorePath` as return type for regular list calls?